types.go

```diff
@@ -0,0 +1,921 @@
+	if rhs, ok := p.(BooleanParameter); ok {
+		return b == rhs
+	}
+	return false
+}
+
+func (b BooleanParameter) ToProto() *proto.Type_Parameter {
+	return &proto.Type_Parameter{Parameter: &proto.Type_Parameter_Boolean{
+		Boolean: bool(b)}}
+}
+
+// IntegerParameter is the type parameter like 10 in VARCHAR<10>
+type IntegerParameter int64
+
+func (b IntegerParameter) Equals(p TypeParam) bool {
+	if rhs, ok := p.(IntegerParameter); ok {
+		return b == rhs
+	}
+	return false
+}
+
+func (p IntegerParameter) ToProto() *proto.Type_Parameter {
+	return &proto.Type_Parameter{Parameter: &proto.Type_Parameter_Integer{
+		Integer: int64(p)}}
+}
+
+// EnumParameter is a type parameter that is some enum value
+type EnumParameter string
+
+func (b EnumParameter) Equals(p TypeParam) bool {
+	if rhs, ok := p.(EnumParameter); ok {
+		return b == rhs
+	}
+	return false
+}
+
+func (p EnumParameter) ToProto() *proto.Type_Parameter {
+	return &proto.Type_Parameter{Parameter: &proto.Type_Parameter_Enum{
+		Enum: string(p)}}
+}
+
+// StringParameter is a type parameter which is a string value
+type StringParameter string
+
+func (b StringParameter) Equals(p TypeParam) bool {
+	if rhs, ok := p.(StringParameter); ok {
+		return b == rhs
+	}
+	return false
+}
+
+func (p StringParameter) ToProto() *proto.Type_Parameter {
+	return &proto.Type_Parameter{Parameter: &proto.Type_Parameter_String_{
+		String_: string(p)}}
+}
+
+// TypeParamFromProto converts a protobuf Type_Parameter message to
+// a TypeParam object for processing.
+func TypeParamFromProto(p *proto.Type_Parameter) TypeParam {
+	switch p := p.Parameter.(type) {
+	case *proto.Type_Parameter_Null:
+		return NullParameter{}
+	case *proto.Type_Parameter_Boolean:
+		return BooleanParameter(p.Boolean)
+	case *proto.Type_Parameter_DataType:
+		return &DataTypeParameter{TypeFromProto(p.DataType)}
+	case *proto.Type_Parameter_Integer:
+		return IntegerParameter(p.Integer)
+	case *proto.Type_Parameter_Enum:
+		return EnumParameter(p.Enum)
+	case *proto.Type_Parameter_String_:
+		return StringParameter(p.String_)
+	}
+	return nil
+}
+
+type UserDefinedType struct {
+	Nullability      Nullability
+	TypeVariationRef uint32
+	TypeReference    uint32
+	TypeParameters   []TypeParam
+}
+
+func (*UserDefinedType) isRootRef()                          {}
+func (s *UserDefinedType) GetType() Type                     { return s }
+func (s *UserDefinedType) GetNullability() Nullability       { return s.Nullability }
+func (s *UserDefinedType) GetTypeVariationReference() uint32 { return s.TypeVariationRef }
+
+func (t *UserDefinedType) Equals(rhs Type) bool {
+	if other, ok := rhs.(*UserDefinedType); ok {
+		switch {
+		case t.Nullability != other.Nullability:
+			return false
+		case t.TypeVariationRef != other.TypeVariationRef:
+			return false
+		case t.TypeReference != other.TypeReference:
+			return false
+		case len(t.TypeParameters) != len(other.TypeParameters):
+			return false
+		}
+
```

Comment:
Do we need a check here comparing len(t.TypeParameters) and len(other.TypeParameters)?

Reply:
we check it on line 816, if they aren't equal, we'll have returned false already.